### ISSUE TITLE
[autoscaler] Service and Ingress resource per worker

### ISSUE THREAD

**Issue by PidgeyBE** [ray-project/ray#9042]
Dear all

I'm trying to create an application which almost entirely runs on Ray. Currently it's running fine on a Ray cluster that I've deployed manually, but I would like to use the ray autoscaler (to pay less for GPU etc ;-)).
The application basically spawns ML pipelines in Ray actors and serves HTTP API servers, RTSP streams, etc.
In order to expose the communication endpoints to the outside world, the ray worker pods are exposed via kubernetes services and these services are exposed via Ingress resources.
My current deployment looks like this:
* A ray head and a few worker pods
* A kubernetes service _for each worker_ exposing port range 7000->7010 on each respective worker
* Ingress rules for each service, exposing the ports like: 
<details>
  <summary>ingress translation</summary>

| ingress rule | service:port |
| ------------- | ------------- |
| https://domainname.com/worker1/7000 | ray-worker-1:7000 | 
| https://domainname.com/worker1/... | ray-worker-1:... | 
| https://domainname.com/worker1/7010 | ray-worker-1:7010 | 
| https://domainname.com/worker2/7000 | ray-worker-2:7000 | 
| https://domainname.com/worker2/... | ray-worker-2:... | 
| https://domainname.com/worker2/7010 | ray-worker-2:7010 | 
| .... | .... |

</details>
When e.g. an HTTP API server is spawned inside an actor, it knows the port it's running on (e.g. 7000) and knows the worker it's running in (currently via the HOSTNAME env var), so based on this information it can determine on which URL the API server is exposed for end users.

I'd like to use the same architecture with autoscale, but I'm missing the possibility to deploy/remove a kubernetes service and ingress resource along with each pod. To accomplish this, the autoscale yaml could be like this:
<details>
  <summary>autoscale.yaml</summary>

```
provider:
    type: kubernetes
# Kubernetes pod config for the head node pod.
head_node:
    pod:
        apiVersion: v1
        kind: Pod
        metadata:
            # Automatically generates a name for the pod with this prefix.
            generateName: ray-head-
            labels:
                component: ray-head
        spec:
    service:  # optional
        # Service that maps to the head node of the Ray cluster.
        apiVersion: v1
        kind: Service
        metadata:
    ingress:  # optional
        apiVersion: extensions/v1beta1
        kind: Ingress
        metadata:
# Kubernetes pod config for worker node pods.
worker_nodes:
    pod:
        apiVersion: v1
        kind: Pod
        metadata:
            # Automatically generates a name for the pod with this prefix.
            generateName: ray-worker-
        labels:
            component: 
    service:  # optional
        # Service that maps to the head node of the Ray cluster.
        apiVersion: v1
        kind: Service
        metadata:
    ingress:  # optional
        apiVersion: extensions/v1beta1
        kind: Ingress
        metadata:
```

</details>

The only thing that's not 100% straightforward is that there should be some intelligence in the autoscaler to map the right service on the right worker. E.g. by assigning a unique label to every pod and using this label as selector for the service, so there is a 1:1 relationship between service and pod.

I hope my goal is clear and I think it'll be useful functionality for other people too.
Thanks for taking this into discussion/consideration!

BR, Pieterjan

CC @edoakes

**Comment by edoakes** [comment]
@thomasdesr any thoughts here?

**Comment by edoakes** [comment]
@PidgeyBE so basically the missing pieces are:

1. Add the ability to specify an optional service and ingress per-pod (instead of globally like we currently do).
2. Add some ability to auto-fill the pod's name to back the created service (with some magic token like `$POD_NAME`.

I think this sounds reasonable to me. This may actually make it simpler for users to deploy services instead of needing to match the label selector manually between the service definition in the service provider spec and the pod definition below.

**Comment by PidgeyBE** [comment]
Indeed that's it!
Be aware that if the changes are done, it will be a breaking change, cause the yaml hierarchy will be different. Or do you want to keep supporting "legacy yamls"?

**Comment by thomasdesr** [comment]
Hey @PidgeyBE!

Thanks this seems super useful. I want to make sure I understand what you're trying to setup you want to switch from:

## old
![image](https://user-images.githubusercontent.com/681004/86025324-0e3d0a80-b9fc-11ea-9924-2ab506525a77.png)


## new
![image](https://user-images.githubusercontent.com/681004/86025393-1f861700-b9fc-11ea-96f3-fe55ff021aa0.png)


If you haven't seen headless services &| stateful sets, they may also be useful as you explore out what the right solution is going to be. :D

Headless services would let you bypass the single-ip nature of a single service and might let you avoid having to create a service-per-pod. Stateful sets would give you sticky & predictable names for each pod.

**Comment by PidgeyBE** [comment]
Hey @thomasdesr ,

That's indeed exactly what I'd like to achieve!
I have no experience with headless services &| stateful sets, so I'll look into them! Thanks a lot!

**Comment by PidgeyBE** [comment]
Hey @thomasdesr 
I read about stateful sets and headless service, but I think, to be able to use that as a solution, we are missing one part.
We are not able to expose the pods in the stateful set via ingress. Coincidentally someone else opened a ticket about that yesterday: https://github.com/kubernetes/kubernetes/issues/92639

**Comment by PidgeyBE** [comment]
I'm digging into the autoscaler code and one fundamental issue I see with statefulsets is that individual nodes/pods can't be scaled down (https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#deployment-and-scaling-guarantees).
The ray autoscaler tries to scale down the idle/inactive workers nodes, but in statefulsets you can't terminate an individual pod of the set. If you scale a statefulset down, it will just remove the one it created last. But this might be a worker node which is not idle!

So this means, statefulsets can't be used in a ray autoscaler context.
So the solution left is:
* Stick to creating nodes (as pods)
* deploy/delete additional service (and ingress) when scaling the nodes

**Comment by PidgeyBE** [comment]
@thomasdesr @edoakes I've made a proof of concept and tested it. It works for me: https://github.com/PidgeyBE/ray/pull/1/files
For each pod that is created/terminated:
* an optional service is created with a selector on a unique label I added to the pods. The service gets the same name as the pod.
* an optional ingress rule is created, which maps automatically on the service that was created in the step before.

Let me know what you think!
